tool/metrics_monitor.py

```diff
@@ -0,0 +1,75 @@
+import time
+import datetime as dt
+import threading
+import sys
+import psutil
+import firebase_admin
+from firebase_admin import credentials
+from firebase_admin import firestore
+from core_data_modules.logging import Logger
+
+COLLECTION = 'pipeline_metrics'
+DEFAULT_INTERVAL = 3
+
+log = Logger(__name__)
+
+
+def initialize_firebase(CRYPTO_TOKEN_PATH):
+    global firebase_client
+    log.info("Setting up Firebase client")
+    firebase_cred = credentials.Certificate(CRYPTO_TOKEN_PATH)
+    firebase_admin.initialize_app(firebase_cred)
+    firebase_client = firestore.client()
+    log.info("Firebase client ready")
+
+
+def get_metrics(interval):
+    while True:
+        metrics = {}
+
+        # record datetime
+        metrics['datetime'] = dt.datetime.now().isoformat() + 'Z'
+
+        # current cpu utlization
+        cpu_utilization = psutil.cpu_percent(interval=0.1)
+        metrics['cpu_percent'] = cpu_utilization
+
+        # cpu load over the last 1, 5 and 15 minutes in percentage
+        cpu_load = [round((value / psutil.cpu_count() * 100), 2)
+                    for value in psutil.getloadavg()]
+        metrics['cpu_load_interval_percent'] = cpu_load
+
+        # memory usage
+        memory_usage = psutil.virtual_memory()
+        metrics['memory_usage'] = dict(
+            {
+                'available': memory_usage[1],
+                'used': memory_usage[3],
+                'percent': memory_usage[2],
+                'free': memory_usage[4]
+            }
+        )
+
+        # disk usage
+        disk_usage = psutil.disk_usage('/')
+        metrics['disk_usage'] = dict(disk_usage._asdict())
+
+        log.info("Recorded metrics: {}".format(metrics))
+
+        publish_metrics_to_firestore(metrics)
+        time.sleep(interval)
+
+
+def publish_metrics_to_firestore(metrics):
+    firebase_client.collection(COLLECTION).add(metrics)
+    log.info("Successfully published metrics to firebase {} collection".format(COLLECTION))
+
+
+def run_metric_monitor(interval=DEFAULT_INTERVAL):
+    initialize_firebase(sys.argv[1])
```

Comment:
We might need to make changes to this so that the firebase initialization happens outside this function call, but we can figure out exactly where and how when integrating this into the pipelines
cc @lukechurch

Reply:
Yeah... @lukechurch  has said he'd know where to refactor in this code into the pipeline